File: SapphireYours_Common/src/grafl/sy/screens/LevelSettingsDialog.java
package grafl.sy.screens;

import android.view.KeyEvent;
import grafl.sy.buttons.BackButton;
import grafl.sy.buttons.ButtonSet;
import grafl.sy.buttons.IncreaseValueButton;
import grafl.sy.buttons.ReduceValueButton;
import grafl.sy.game.Game;
import grafl.sy.logic.Level;
import grafl.sy.renderer.TextRenderer;
import grafl.sy.renderer.VectorRenderer;

public class LevelSettingsDialog extends Screen
{
	Level level;
	int totalloot;

	int dialogx;
	int dialogy;
	int dialogwidth;
	int dialogheight;
	
	ButtonSet  buttons;
	int selectedline;

	public LevelSettingsDialog(Game game, Level level)
	{
		super(game);
		this.level = level;
		this.totalloot = level.calculateMaximumLoot(false);
		this.buttons = new ButtonSet(false);		
		this.selectedline = -1;						
	}
	
	@Override
	public void resize(int width, int height)
	{
		super.resize(width,height);
		drawOrLayout(false);
	}

	@Override
	public boolean isOverlay()
	{	return true;
	}
	


	public void draw()
	{
		drawOrLayout(true);
	}
	
	private void drawOrLayout(boolean draw)
	{
		float scaling = game.detailScale;		
		float th = 27*scaling;
		float lspace = 5*scaling; 
		int border = (int) (15*scaling);				
		 
		VectorRenderer vr = game.vectorRenderer;				
		TextRenderer tr = game.textRenderer;				

		// layout depends on menu width which depends on the screen size
		dialogwidth = Math.min(300, screenwidth);					
		dialogx = (screenwidth-dialogwidth)/2;
		int col = 0xff000000;
		int bgcolor = 0xffbbbbbb;
		int labelx = dialogx + border;
		int valuex = dialogx+dialogwidth-dialogwidth/3;
		int valuewidth = dialogwidth/3-20;
		int buttonx1 = valuex - dialogwidth/4;
		int buttonx2 = valuex + dialogwidth/4;
		int y = dialogy+border;	 // when painting from top to bottom increase y position
				
		// at first call or after size change, initialize buttons
		if (!draw)		
		{	
			buttons = new ButtonSet(false);
			selectedline = game.usingKeyboardInput ? 0: -1;
		}
		// when in drawing mode, initialize renderers and draw background and all buttons
		else
		{	vr.startDrawing (screenwidth, screenheight);
			tr.startDrawing (screenwidth, screenheight);		
			vr.addRoundedRect(dialogx,dialogy,dialogwidth,dialogheight, 20*scaling, 20*scaling+1, bgcolor);
			buttons.draw(vr);
		}

		// add caption
		if (draw)
		{	tr.addString ("Level Settings", labelx, y, th, false, col, TextRenderer.WEIGHT_PLAIN);
		}
		y += th + 2*lspace;
		

		// add difficulty text
		int difficultyliney = y - dialogy;
		if (draw)
		{
			tr.addString ("Difficulty", labelx, y, th, false, col, TextRenderer.WEIGHT_PLAIN);
			String difname = Game.getNameForDifficulty(level.getDifficulty());
			float tw = tr.determineStringWidth(difname, th);
			int c = col;
			if (selectedline==0)
			{	c = bgcolor;
				vr.addRectangle(valuex-valuewidth/2,y,valuewidth,th, col);
			}
			tr.addString (difname, valuex-tw/2,y, th, false, c, TextRenderer.WEIGHT_BOLD);
		}		
		y += th + lspace;
		
		// add category name
		int categoryliney = y - dialogy;
		if (draw)
		{
			tr.addString ("Category", labelx, y, th, false, col, TextRenderer.WEIGHT_PLAIN);
			String catname = Game.getNameForCategory(level.getCategory());
			float tw = tr.determineStringWidth(catname, th);
			int c = col;
			if (selectedline==1)
			{	c = bgcolor;
				vr.addRectangle(valuex-valuewidth/2,y,valuewidth,th, col);
			}
			tr.addString (catname, valuex-tw/2,y, th, false, c, TextRenderer.WEIGHT_BOLD);
		}		
		y += th +lspace;
		
		// add loot speed
		int lootliney = y - dialogy;
		if (draw)
		{
			tr.addString ("Collect", labelx, y, th, false, col, TextRenderer.WEIGHT_PLAIN);
			String ltxt;
			int l = level.getLoot();
			if (l<totalloot)
			{	ltxt = "- " + (totalloot-l);
			}
			else if (l>totalloot)
			{	ltxt = "+ " + (l-totalloot);
			}
			else
			{	ltxt = "All";
			}
			float tw = tr.determineStringWidth(ltxt, th);
			int c = col;
			if (selectedline==2)
			{	c = bgcolor;
				vr.addRectangle(valuex-valuewidth/2,y,valuewidth,th, col);
			}
			tr.addString (ltxt, valuex-tw/2,y, th, false, c, TextRenderer.WEIGHT_BOLD);
		}		
		y += th + lspace;

		// add swamp speed
		int swamprateliney = y - dialogy;
		if (draw)
		{
			tr.addString ("Swamp", labelx, y, th, false, col, TextRenderer.WEIGHT_PLAIN);
			String sptxt = ""+level.getSwampRate();
			float tw = tr.determineStringWidth(sptxt, th);
			int c = col;
			if (selectedline==3)
			{	c = bgcolor;
				vr.addRectangle(valuex-valuewidth/2,y,valuewidth,th, col);
			}
			tr.addString (sptxt, valuex-tw/2,y, th, false, c, TextRenderer.WEIGHT_BOLD);
		}		
		y += th + lspace;
		
		
		
		if (!draw)
		{
			// compute the dialog size for subsequent paints
			y+= border;
			dialogheight = y - dialogy;
			dialogy = screenheight/2 - dialogheight/2; 		
			
			// add buttons at correct y positions
			buttons.add(
				new ReduceValueButton(game, buttonx1, dialogy+difficultyliney, th,th,
				 	new Runnable(){public void run(){selectedline=-1; reduceDifficulty();}}
			));
			buttons.add(
				new IncreaseValueButton(game, buttonx2-th, dialogy+difficultyliney, th,th, 
					new Runnable(){public void run(){selectedline=-1;incrementDifficulty();}}
			));

			buttons.add(
				new ReduceValueButton(game, buttonx1, dialogy+categoryliney, th,th,
				 	new Runnable(){public void run(){selectedline=-1;reduceCategory();}}
			));
			buttons.add(
				new IncreaseValueButton(game, buttonx2-th, dialogy+categoryliney, th,th, 
					new Runnable(){public void run(){selectedline=-1;incrementCategory();}}
			));
			
			buttons.add(
				new ReduceValueButton(game, buttonx1, dialogy+lootliney, th,th,
				 	new Runnable(){public void run(){selectedline=-1;reduceLoot();}}
			));
			buttons.add(
				new IncreaseValueButton(game, buttonx2-th, dialogy+lootliney, th,th, 
					new Runnable(){public void run(){selectedline=-1;incrementLoot();}}
			));
			
			buttons.add(
				new ReduceValueButton(game, buttonx1, dialogy+swamprateliney, th,th,
				 	new Runnable(){public void run(){selectedline=-1;reduceSwampRate();}}
			));
			buttons.add(
				new IncreaseValueButton(game, buttonx2-th, dialogy+swamprateliney, th,th, 
					new Runnable(){public void run(){selectedline=-1;incrementSwampRate();}}
			));
			
			// add exit button
			buttons.add(
				new BackButton(game, dialogx+dialogwidth-border-th, dialogy+border, th,th,
				 	new Runnable(){public void run(){ game.removeScreen();}}
			));			
		}
		else
		{	// flush everything to screen
			vr.flush();
			tr.flush();
		}
	}
	
    @Override
	public void onPointerDown(int x, int y)
	{	
		buttons.onPointerDown(x,y);		
	}
    @Override
	public void onPointerUp()
	{
		buttons.onPointerUp();		
	}
    @Override
	public void onPointerMove(int x, int y)
	{	
		buttons.onPointerMove(x,y);		
	}
	@Override	
    public void handleKeyEvent(KeyEvent event)
    {
    	if (event.getAction()==KeyEvent.ACTION_DOWN)
		{
			switch (event.getKeyCode())
			{	case KeyEvent.KEYCODE_DPAD_UP:
					selectedline = Math.max(selectedline-1,0);
					break;
				case KeyEvent.KEYCODE_DPAD_DOWN:
					selectedline = Math.min(selectedline+1,3);
					break;
				case KeyEvent.KEYCODE_DPAD_LEFT:
					switch (selectedline)
					{	case 0: reduceDifficulty(); 
								break;
						case 1: reduceCategory(); 
								break;
						case 2: reduceLoot(); 
								break;
						case 3: reduceSwampRate(); 
								break;
						default: selectedline=0;
								break;
					}
					break;
				case KeyEvent.KEYCODE_DPAD_RIGHT:
					switch (selectedline)
					{	case 0: incrementDifficulty(); 
								break;
						case 1: incrementCategory(); 
								break;
						case 2: incrementLoot(); 
								break;
						case 3: incrementSwampRate(); 
								break;
						default: selectedline=0;
								break;
					}
					break;
				case KeyEvent.KEYCODE_ENTER:			
					game.removeScreen();
					break;
			}
		}
    
		// buttons.handleKeyEvent(event);				    
    }
	
	private void reduceDifficulty()
	{
		level.setDifficulty(Math.max(level.getDifficulty()-1,1));		
	}
	private void incrementDifficulty()
	{
		level.setDifficulty(Math.min(level.getDifficulty()+1,9));			
	}
	
	private void reduceCategory()
	{
		level.setCategory(Math.max(level.getCategory()-1,0));		
	}
	private void incrementCategory()
	{
		level.setCategory(Math.min(level.getCategory()+1,6));			
	}
	
	private void reduceLoot()
	{
		level.setLoot(Math.max(level.getLoot()-1,0));		
	}
	private void incrementLoot()
	{
		level.setLoot(level.getLoot()+1);			
	}

	private void reduceSwampRate()
	{
		level.setSwampRate(Math.max(level.getSwampRate()-1,0));		
	}
	private void incrementSwampRate()
	{
		level.setSwampRate(level.getSwampRate()+1);			
	}
	
	
}

File: game/src/screen/LevelSettingsDialog.js
"use strict";
var LevelSettingsDialog = function()
{   Screen.call(this);
    
    this.level = null;  
    this.selected = 0;
    this.pointerx = 0;
    this.pointery = 0;
};
LevelSettingsDialog.prototype = Object.create(Screen.prototype);
    
    
LevelSettingsDialog.prototype.$ = function(game, level)
{   Screen.prototype.$.call(this,game); 

    this.level = level;
    this.selected = -1;
    return this;
};

LevelSettingsDialog.prototype.isOverlay = function()
{   return true;
};

LevelSettingsDialog.prototype.draw = function()
{
    var vr = this.game.vectorRenderer;                
    var tr = this.game.textRenderer;                
    vr.startDrawing ();
    tr.startDrawing ();    
    
    var w = Math.min(370, this.game.screenwidth-20);
    var h = 435;
    var x = 10;
    var y = this.game.screenheight-h-10;
    var spacing = 70;
    
    vr.addRoundedRect (x,y,w,h, 4,6, 0xcc000000);
    
    this.renderSetting(x+20,y+10, 
        "Difficulty", Game.getNameForDifficulty(this.level.difficulty),
        this.selected==0 );    
    this.renderSetting(x+20,y+10+spacing, 
        "Category", Game.getNameForCategory(this.level.category),
        this.selected==1 );
    this.renderSetting(x+20,y+10+spacing*2, 
        "Gems points needed", ""+this.level.loot,
        this.selected==2 );
    this.renderSetting(x+20,y+10+spacing*3, 
        "Swamp spreading speed", this.level.swamprate ? ""+this.level.swamprate : "Off",
        this.selected==3 );
    this.renderSetting(x+20,y+10+spacing*4, 
        "Robot speed", this.level.robotspeed ? ""+this.level.robotspeed : "Off",
        this.selected==4 );
            
    this.renderSetting(x+20,y+10+spacing*5, "", "Done", this.selected==5 );
            
    vr.flush();
    tr.flush();    
};

LevelSettingsDialog.prototype.renderSetting = function(x,y,label,value,highlight)
{
    this.game.textRenderer.addString(label, x,y, 20, false, 0xff888888, TextRenderer.WEIGHT_PLAIN);
    
    var fcol = 0xffdddddd;
    var bcol = 0xff393939;
    if (highlight)
    {   var fcol = 0xff393939;
        var bcol = 0xffdddddd;
    }
    this.game.vectorRenderer.addRoundedRect(x,y+25,330,40, 4,6, bcol);   
    this.game.textRenderer.addString(value, x+20,y+35, 20, false, fcol, TextRenderer.WEIGHT_BOLD);        
    if (label.length>0)
    {   this.game.textRenderer.addString("-", x+240,y+30, 30, false, fcol, TextRenderer.WEIGHT_BOLD);    
        this.game.textRenderer.addString("+", x+280,y+33, 30, false, fcol, TextRenderer.WEIGHT_BOLD);    
        
    }
};

LevelSettingsDialog.prototype.findAction = function(px, py)
{    
    var h = 435;
    var spacing = 70;
    for (var i=0; i<6; i++)
    {   var x = 10+20;
        var y = this.game.screenheight-h-10+10+spacing*i+25;
        if (px>=x && px<x+330 && py>=y && py<y+40) { return i; }        
    }
    return -1;
};

LevelSettingsDialog.prototype.onBackNavigation = function()
{
    this.game.removeScreen();
    this.game.getTopScreen().afterScreenCreation();
};
        
LevelSettingsDialog.prototype.onKeyDown = function(code)
{   
    var l = this.level;
    switch (code)
    {   case KeyEvent.UP:
        {   if (this.selected>0)
            {   this.selected--;
                this.setDirty();
            }
            break;                  
        }
        case KeyEvent.DOWN:
        {   if (this.selected<5)
            {   this.selected++;
                this.setDirty();
            }
        }
        case KeyEvent.LEFT:
        {   switch(this.selected)
            {   case 0: // difficulty
                {   if (l.difficulty>2) { l.difficulty--; }
                    this.setDirty();
                    break;
                }            
                case 1: // category
                {   if (l.category>0) { l.category--; }
                    this.setDirty();
                    break;
                }                        
                case 2: // loot
                {   if (l.loot>0) { l.loot--; }
                    this.setDirty();
                    break;
                }                        
                case 3: // swamp speed
                {   if (l.swamprate>0) { l.swamprate--; }
                    this.setDirty();
                    break;
                }                        
                case 4: // robot speed
                {   if (l.robotspeed>0) { l.robotspeed--; }
                    this.setDirty();
                    break;
                }                        
            }
            break;
        }
        case KeyEvent.RIGHT:
        {   switch(this.selected)
            {   case 0: // difficulty
                {   if (l.difficulty<9) { l.difficulty++; }
                    this.setDirty();
                    break;
                }            
                case 1: // category
                {   if (l.category<7) { l.category++; }
                    this.setDirty();
                    break;
                }                        
                case 2: // loot
                {   l.loot++;
                    this.setDirty();
                    break;
                }                        
                case 3: // swamp speed
                {   l.swamprate++;
                    this.setDirty();
                    break;
                }                        
                case 4: // robot speed
                {   l.robotspeed++;
                    this.setDirty();
                    break;
                }                        
            }
            break;
        }
        case KeyEvent.A:
        {   if (this.selected==5) 
            {   this.onBackNavigation();
            }
            break;
        }
    }
};


LevelSettingsDialog.prototype.onPointerDown = function(x, y)
{
    this.pointerx = x;
    this.pointery = y;
    
    var sel = this.findAction(x,y);
    if (sel!=this.selected)
    {   this.selected = sel;
        this.setDirty();
    }
};
    
LevelSettingsDialog.prototype.onPointerUp = function()
{
    if (this.selected>=0 && this.findAction(this.pointerx,this.pointery)==this.selected)
    {           
        if (this.selected<=4)
        {   this.onKeyDown(this.pointerx < 295 ? KeyEvent.LEFT : KeyEvent.RIGHT);
        } 
        else
        {   this.onKeyDown(KeyEvent.A);
        }
    }       
};
    
LevelSettingsDialog.prototype.onPointerMove = function(x, y)
{
    this.pointerx = x;
    this.pointery = y;
};


